engine/src/cmd/damageable.h

```diff
@@ -26,78 +26,142 @@
+    Damageable(): hull(&layers[0]),
+        armor(&layers[1]),
+        shield(&layers[2]),
+        current_hull(&hull->facets[0].health),
+        max_hull(&hull->facets[0].factory_max_health),
+        dying(false),
+        killed(false) {}
 
 protected:
-  virtual ~Damageable() = default;
-  // forbidden
-  Damageable( const Damageable& ) = delete;
-  // forbidden
-  Damageable& operator=( const Damageable& ) = delete;
+    virtual ~Damageable() = default;
+    // forbidden
+    Damageable( const Damageable& ) = delete;
+    // forbidden
+    Damageable& operator=( const Damageable& ) = delete;
 
 public:
-  float GetHull() const { return hull;}
-
-  void ArmorData( float armor[8] ) const;
-
-  // Is the shield up from this position
-  bool ShieldUp( const Vector& ) const;
-
-  //applies damage from the given pnt to the shield, and returns % damage applied and applies lighitn
-  virtual float DealDamageToShield( const Vector &pnt, float &Damage );
-
-  float FShieldData() const;
-  float RShieldData() const;
-  float LShieldData() const;
-  float BShieldData() const;
-//short fix
-//  void ArmorData( float armor[8] ) const;
-//Gets the current status of the hull
-//  float GetHull() const
-//  {
-//      return hull;
-//  }
-  float GetHullPercent() const
-  {
-      return maxhull != 0 ? hull/maxhull : hull;
-  }
-
-  //reduces shields to X percentage and reduces shield recharge to Y percentage
-  void leach( float XshieldPercent, float YrechargePercent, float ZenergyPercent );
-
-  virtual float DealDamageToHull( const Vector &pnt, float Damage );
-
-  float MaxShieldVal() const;
-//regenerates all 2,4, or 6 shields for 1 SIMULATION_ATOM
-  virtual void RegenShields() = 0;
-  bool withinShield( const ShieldFacing &facing, float theta, float rho );
-
-  static float totalShieldVal( const Shield &shield );
-  static float currentTotalShieldVal( const Shield &shield );
-  static float totalShieldEnergyCapacitance( const Shield &shield );
+    // We follow the existing convention of GetX for the actual health value
+    // because we are constrained by existing python interfaces, which cannot
+    // be easily changed.
+    const float GetHull() const {
+        return hull->facets[0].health;
+    }
+
+    const float GetArmor(int facet = 0) const {
+        return armor->facets[facet].health;
+    }
+
+    const float GetShield(int facet = 0) const {
+        return shield->facets[facet].health;
+    }
+
+    DamageableLayer& GetHullLayer() {
+        return layers[0];
+    }
+
+    DamageableLayer& GetArmorLayer() {
+        return layers[1];
+    }
+
+    DamageableLayer& GetShieldLayer() {
+        return layers[2];
+    }
+
+    const float GetShieldRegeneration() const {
+        return shield->facets[0].regeneration;
+    }
+
+    const float GetHullPercent() const
+    {
+        return hull->facets[0].Percent();
+    }
+
+    const float GetShieldPercent() const
+    {
+        return shield->facets[0].Percent();
+    }
```

Comment:
The above functions should probably be declared inline.

Reply:
It's short and it's in the header but I'm not convinced it should be inlined. I'd prefer someone optimizes things after profiling them. In any case, the low hanging fruit is introducing threads.

Reply:
functions definitions in headers should be inline; if they shouldn't be, then they don't belong in headers.